// @progress/kendo-react-dateinputs/dist/npm/timepicker/services/DOMService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DOMService = void 0;
var utils_1 = require("../../utils");
var kendo_react_common_1 = require("@progress/kendo-react-common");
/**
 * @hidden
 */
var DOMService = /** @class */ (function () {
    function DOMService() {
        this.itemHeight = 0;
        this.timeListHeight = 0;
        this.didCalculate = false;
    }
    DOMService.prototype.ensureHeights = function () {
        if (this.timeListHeight !== undefined) {
            return;
        }
        this.calculateHeights();
    };
    DOMService.prototype.calculateHeights = function (container) {
        if (!kendo_react_common_1.canUseDOM) {
            return;
        }
        var div = (0, utils_1.domContainerFactory)('div');
        var span = (0, utils_1.domContainerFactory)('span');
        var ul = (0, utils_1.domContainerFactory)('ul');
        var li = (0, utils_1.domContainerFactory)('li');
        var listItem = function () { return li('<span>02</span>', 'k-item'); };
        var list = function () { return ul([listItem()], 'k-reset'); };
        var scrollable = function () { return (div([list()], 'k-content k-calendar-content k-scrollable k-time-container')); };
        var timeListWrapper = function () {
            return div([
                span('minute', 'k-title'),
                div([scrollable()], 'k-time-list')
            ], 'k-time-list-wrapper', { left: '-10000px', position: 'absolute' });
        };
        var timeWrapper = timeListWrapper();
        var listContainer = container && container.querySelector('.k-time-container');
        var hostContainer = listContainer || document.body;
        var wrapper = hostContainer.appendChild(timeWrapper);
        this.timeListHeight = wrapper.querySelector('.k-scrollable').offsetHeight;
        this.itemHeight = wrapper.querySelector('li').offsetHeight;
        hostContainer.removeChild(wrapper);
        this.didCalculate = true;
    };
    return DOMService;
}());
exports.DOMService = DOMService;
