// @progress/kendo-react-labels/dist/npm/Hint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hint = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var kendo_react_common_2 = require("@progress/kendo-react-common");
var package_metadata_1 = require("./package-metadata");
/**
 * Represents the KendoReact Hint component.
 * Render the hint text that will be shown underneath the form editor.
 */
var Hint = function (props) {
    (0, kendo_react_common_2.validatePackage)(package_metadata_1.packageMetadata);
    var hintClassName = (0, kendo_react_common_1.classNames)({
        'k-form-hint': true,
        'k-text-start': props.direction === 'start',
        'k-text-end': props.direction === 'end',
        'k-text-disabled': props.editorDisabled === true
    }, props.className);
    return (React.createElement("div", { id: props.id, style: props.style, className: hintClassName }, props.children));
};
exports.Hint = Hint;
exports.Hint.propTypes = {
    id: PropTypes.string,
    direction: PropTypes.oneOf(['start', 'end']),
    children: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.node
    ]),
    style: PropTypes.object,
    className: PropTypes.string,
    editorDisabled: PropTypes.bool
};
exports.Hint.defaultProps = {
    direction: 'start'
};
exports.Hint.displayName = 'KendoReactHint';
