// @progress/kendo-react-data-tools/dist/npm/filter/Group.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Group = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var kendo_data_query_1 = require("@progress/kendo-data-query");
var kendo_react_buttons_1 = require("@progress/kendo-react-buttons");
var kendo_react_intl_1 = require("@progress/kendo-react-intl");
var Expression_1 = require("./Expression");
var operators_1 = require("./operators");
var messages_1 = require("../messages");
/**
 * @hidden
 */
var Group = /** @class */ (function (_super) {
    __extends(Group, _super);
    function Group() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onGroupRemove = function (event) {
            var ev = {
                filter: _this.props.filter,
                target: _this,
                syntheticEvent: event,
                nativeEvent: event.nativeEvent
            };
            _this.props.onRemove.call(undefined, ev);
        };
        _this.onAddExpression = function (event) {
            var prevFilter = _this.props.filter;
            var firstField = _this.props.fields[0];
            var newExpression = { field: firstField.name, operator: firstField.operators[0].operator };
            newExpression.value = (0, operators_1.stringOperator)(newExpression.operator) ? '' : null;
            var ev = {
                nextFilter: __assign(__assign({}, prevFilter), { filters: __spreadArray(__spreadArray([], prevFilter.filters, true), [newExpression], false) }),
                prevFilter: prevFilter,
                target: _this,
                syntheticEvent: event,
                nativeEvent: event.nativeEvent
            };
            _this.props.onChange.call(undefined, ev);
        };
        _this.onAddGroup = function (event) {
            var prevFilter = _this.props.filter;
            var ev = {
                nextFilter: __assign(__assign({}, prevFilter), { filters: __spreadArray(__spreadArray([], prevFilter.filters, true), [__assign({}, _this.props.defaultGroupFilter)], false) }),
                prevFilter: prevFilter,
                target: _this,
                syntheticEvent: event,
                nativeEvent: event.nativeEvent
            };
            _this.props.onChange.call(undefined, ev);
        };
        _this.onLogicAnd = function (event) { return _this.changeLogic('and', event); };
        _this.onLogicOr = function (event) { return _this.changeLogic('or', event); };
        _this.changeLogic = function (logic, event) {
            var prevFilter = _this.props.filter;
            if (prevFilter.logic !== logic) {
                var ev = {
                    nextFilter: __assign(__assign({}, prevFilter), { logic: logic }),
                    prevFilter: prevFilter,
                    target: _this,
                    syntheticEvent: event,
                    nativeEvent: event.nativeEvent
                };
                _this.props.onChange.call(undefined, ev);
            }
        };
        _this.replaceFilter = function (prevFilter, nextFilter) {
            var filter = _this.props.filter;
            var filters = filter.filters.map(function (f) { return f === prevFilter ? nextFilter : f; });
            return __assign(__assign({}, filter), { filters: filters });
        };
        _this.onChange = function (event) {
            var nextFilter = _this.replaceFilter(event.prevFilter, event.nextFilter);
            var changeEvent = {
                nextFilter: nextFilter,
                prevFilter: _this.props.filter,
                syntheticEvent: event.syntheticEvent,
                nativeEvent: event.nativeEvent,
                target: _this
            };
            _this.props.onChange.call(undefined, changeEvent);
        };
        _this.onRemove = function (event) {
            var prevFilter = _this.props.filter;
            var filters = prevFilter.filters.filter(function (f) { return f !== event.filter; });
            var changeEvent = {
                nextFilter: __assign(__assign({}, prevFilter), { filters: filters }),
                prevFilter: prevFilter,
                syntheticEvent: event.syntheticEvent,
                nativeEvent: event.nativeEvent,
                target: _this
            };
            _this.props.onChange.call(undefined, changeEvent);
        };
        return _this;
    }
    Group.prototype.render = function () {
        var _this = this;
        var locService = (0, kendo_react_intl_1.provideLocalizationService)(this);
        var _a = this.props, fields = _a.fields, filter = _a.filter, _b = _a.ariaLabel, ariaLabel = _b === void 0 ? locService.toLanguageString(messages_1.filterGroupAriaLabel, messages_1.messages[messages_1.filterGroupAriaLabel]) : _b;
        return (React.createElement(React.Fragment, null,
            React.createElement("div", { className: "k-filter-toolbar", role: "toolbar", "aria-label": ariaLabel },
                React.createElement(kendo_react_buttons_1.Toolbar, { keyboardNavigation: false, role: "" },
                    React.createElement(kendo_react_buttons_1.ToolbarItem, null,
                        React.createElement(kendo_react_buttons_1.ButtonGroup, null,
                            React.createElement(kendo_react_buttons_1.Button, { togglable: true, onClick: this.onLogicAnd, selected: filter.logic === 'and', type: "button" }, locService.toLanguageString(messages_1.filterAndLogic, messages_1.messages[messages_1.filterAndLogic])),
                            React.createElement(kendo_react_buttons_1.Button, { togglable: true, onClick: this.onLogicOr, selected: filter.logic === 'or', type: "button" }, locService.toLanguageString(messages_1.filterOrLogic, messages_1.messages[messages_1.filterOrLogic])))),
                    React.createElement(kendo_react_buttons_1.ToolbarItem, null,
                        React.createElement(kendo_react_buttons_1.Button, { title: locService.toLanguageString(messages_1.filterAddExpression, messages_1.messages[messages_1.filterAddExpression]), icon: "filter-add-expression", type: "button", onClick: this.onAddExpression })),
                    React.createElement(kendo_react_buttons_1.ToolbarItem, null,
                        React.createElement(kendo_react_buttons_1.Button, { title: locService.toLanguageString(messages_1.filterAddGroup, messages_1.messages[messages_1.filterAddGroup]), icon: "filter-add-group", type: "button", onClick: this.onAddGroup })),
                    React.createElement(kendo_react_buttons_1.ToolbarItem, null,
                        React.createElement(kendo_react_buttons_1.Button, { title: locService.toLanguageString(messages_1.filterClose, messages_1.messages[messages_1.filterClose]), icon: "close", fillMode: "flat", type: "button", onClick: this.onGroupRemove })))),
            filter.filters.length > 0 && (React.createElement("ul", { className: "k-filter-lines" }, filter.filters.map(function (f, idx) {
                return (React.createElement("li", { key: idx, className: "k-filter-item" }, (0, kendo_data_query_1.isCompositeFilterDescriptor)(f) ?
                    React.createElement(Group, { filter: f, fields: fields, onChange: _this.onChange, onRemove: _this.onRemove, defaultGroupFilter: _this.props.defaultGroupFilter, ariaLabel: _this.props.ariaLabel, ariaLabelExpression: _this.props.ariaLabelExpression })
                    :
                        React.createElement(Expression_1.Expression, { filter: f, fields: fields, ariaLabel: _this.props.ariaLabelExpression, onChange: _this.onChange, onRemove: _this.onRemove })));
            })))));
    };
    Group.propTypes = {
        filter: PropTypes.object.isRequired,
        fields: PropTypes.array.isRequired,
        ariaLabel: PropTypes.string,
        ariaLabelExpression: PropTypes.string,
        onChange: PropTypes.func.isRequired,
        onRemove: PropTypes.func.isRequired,
        defaultGroupFilter: PropTypes.object.isRequired
    };
    return Group;
}(React.Component));
exports.Group = Group;
(0, kendo_react_intl_1.registerForLocalization)(Group);
