// @progress/kendo-react-inputs/dist/npm/rating/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCorrectValue = exports.calcIsFirstHalf = exports.isSelected = exports.isHalf = exports.getRemainder = exports.toRound = void 0;
/**
 * @hidden
 */
var toRound = function (num, base) {
    if (num % 1) {
        var percent = Math.pow(10, (base.toString().split('.')[1] || base.toString().split('.')[0]).length);
        return Math.round(num * percent) / percent;
    }
    return num;
};
exports.toRound = toRound;
// Opting out of JS `modulo` calculation due to floating-point problems in JS
// We're looking for the reminder - R.
// x / y = a + R
// R = (x / y) - a | *y
// R = x - (y * a)
// : a = Math.floor(x /y );
/**
 * @hidden
 */
var getRemainder = function (x, y) { return x - (y * Math.floor(x / y)); };
exports.getRemainder = getRemainder;
/**
 * @hidden
 */
var isHalf = function (index, value, step) {
    return ((0, exports.toRound)(index - step, step) < value) && (value < index);
};
exports.isHalf = isHalf;
/**
 * @hidden
 */
var isSelected = function (index, value, step, selection) {
    if (value === null) {
        return false;
    }
    if (selection === 'continues') {
        return (0, exports.isHalf)(index, value, step)
            ? (0, exports.toRound)(index - step / 2, step / 2) <= value
            : index <= value;
    }
    if (selection === 'single') {
        return (0, exports.isHalf)(index, value, step)
            ? (0, exports.toRound)(index - step / 2, step / 2) === value
            : index === value;
    }
    return false;
};
exports.isSelected = isSelected;
/**
 * @hidden
 */
var calcIsFirstHalf = function (dir, rect, clientX) {
    return dir === 'ltr'
        ? clientX < (rect.x + (rect.width / 2))
        : clientX > (rect.x + (rect.width / 2));
};
exports.calcIsFirstHalf = calcIsFirstHalf;
/**
 * @hidden
 */
var isCorrectValue = function (min, max, step, value) {
    for (var i = min; i <= max; i += step) {
        if (value === i) {
            return true;
        }
    }
    return false;
};
exports.isCorrectValue = isCorrectValue;
