// @progress/kendo-react-buttons/dist/npm/Chip/ChipList.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChipList = exports.ChipListDataContext = exports.ChipListFocusContext = exports.ChipListSelectionContext = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var selection_reducer_1 = require("./selection-reducer");
var focus_reducer_1 = require("./focus-reducer");
var data_reducer_1 = require("./data-reducer");
var Chip_1 = require("./Chip");
var kendo_react_common_2 = require("@progress/kendo-react-common");
var package_metadata_1 = require("../package-metadata");
/**
 * @hidden
 */
exports.ChipListSelectionContext = React.createContext([null, function (_args) {
        /**/
    }]);
/**
 * @hidden
 */
exports.ChipListFocusContext = React.createContext([null, function (_args) {
        /**/
    }]);
/**
 * @hidden
 */
exports.ChipListDataContext = React.createContext([null, function (_args) {
        /**/
    }]);
/**
 * @hidden
 */
var useSelection = function (defaultValue, args, callback) {
    var _a = React.useState(defaultValue), state = _a[0], setState = _a[1];
    var handleDispatchSelection = function (action) {
        var newState = (0, selection_reducer_1.selectionReducer)(args.state || state, __assign(__assign({}, action), args));
        if (callback) {
            callback(newState, action.event);
        }
        setState(newState);
    };
    return [state, handleDispatchSelection];
};
/**
 * @hidden
 */
var useFocus = function (args) {
    var _a = React.useState(null), state = _a[0], setState = _a[1];
    var handleDispatchFocus = function (action) {
        var newState = (0, focus_reducer_1.focusReducer)(action.payload, __assign(__assign({}, action), args));
        setState(newState);
    };
    return [state, handleDispatchFocus];
};
/**
 * @hidden
 */
var useData = function (defaultData, args, callback) {
    var _a = React.useState(defaultData), state = _a[0], setState = _a[1];
    var handleDispatchData = function (action) {
        var newState = (0, data_reducer_1.dataReducer)(args.state || state, __assign(__assign({}, action), args));
        if (callback) {
            callback(newState, action.event);
        }
        setState(newState);
    };
    return [state, handleDispatchData];
};
/**
 * Represents the ChipList component.
 */
exports.ChipList = React.forwardRef(function (props, ref) {
    var _a;
    (0, kendo_react_common_2.validatePackage)(package_metadata_1.packageMetadata);
    var target = React.useRef(null);
    var chipListRef = React.useRef(null);
    var dir = (0, kendo_react_common_1.useDir)(chipListRef, props.dir);
    var chip = props.chip, id = props.id, style = props.style, tabIndex = props.tabIndex, disabled = props.disabled, size = props.size, className = props.className, ariaDescribedBy = props.ariaDescribedBy, ariaLabelledBy = props.ariaLabelledBy, _b = props.selection, selection = _b === void 0 ? defaultProps.selection : _b, _c = props.value, value = _c === void 0 ? defaultProps.defaultValue : _c, _d = props.defaultData, defaultData = _d === void 0 ? defaultProps.defaultData : _d, _e = props.valueField, valueField = _e === void 0 ? defaultProps.valueField : _e, _f = props.textField, textField = _f === void 0 ? defaultProps.textField : _f, onChange = props.onChange, onDataChange = props.onDataChange;
    var ChipComponent = React.useMemo(function () { return chip || Chip_1.Chip; }, [chip, Chip_1.Chip]);
    React.useImperativeHandle(target, function () { return ({
        element: chipListRef.current,
        props: props
    }); });
    React.useImperativeHandle(ref, function () { return target.current; });
    var handleChange = React.useCallback(function (newValue, event) {
        if (onChange && target.current) {
            onChange.call(undefined, {
                value: newValue,
                target: target.current,
                syntheticEvent: event
            });
        }
    }, [onChange]);
    var _g = useSelection(value, {
        selection: selection,
        state: value
    }, handleChange), stateValue = _g[0], dispatchStateValue = _g[1];
    var handleDataChange = React.useCallback(function (newData, event) {
        if (onDataChange && target.current) {
            onDataChange.call(undefined, {
                value: newData,
                target: target.current,
                syntheticEvent: event
            });
        }
    }, [onDataChange]);
    var _h = useData(props.data || defaultData, {
        state: props.data,
        valueField: valueField
    }, handleDataChange), stateData = _h[0], dispatchData = _h[1];
    var itemsReducer = React.useCallback(function (acc, current) {
        acc.push(current[valueField]);
        return acc;
    }, [valueField]);
    var data = React.useMemo(function () { return props.data || stateData; }, [props.data, stateData]);
    var chipValue = React.useMemo(function () { return value || stateValue; }, [value, stateValue]);
    var items = React.useMemo(function () { return data.reduce(itemsReducer, []); }, [data, itemsReducer]);
    var valueGetter = React.useCallback(function (item) { return (0, kendo_react_common_1.getter)(valueField)(item); }, [valueField]);
    var textGetter = React.useCallback(function (item) { return (0, kendo_react_common_1.getter)(textField)(item); }, [textField]);
    var _j = useFocus({ items: items }), focus = _j[0], dispatchFocus = _j[1];
    var mouseProps = (0, kendo_react_common_1.useMouse)(props, target);
    return (React.createElement(exports.ChipListSelectionContext.Provider, { value: [chipValue, dispatchStateValue] },
        React.createElement(exports.ChipListFocusContext.Provider, { value: [focus, dispatchFocus] },
            React.createElement(exports.ChipListDataContext.Provider, { value: [data, dispatchData] },
                React.createElement("div", __assign({ ref: chipListRef }, mouseProps, { role: 'listbox', id: id, dir: dir, style: style, tabIndex: (0, kendo_react_common_1.getTabIndex)(tabIndex, disabled, undefined), className: (0, kendo_react_common_1.classNames)('k-chip-list', (_a = {
                            'k-rtl': dir === 'rtl',
                            'k-disabled': disabled
                        },
                        _a["k-chip-list-".concat(kendo_react_common_1.kendoThemeMaps.sizeMap[size] || size)] = size,
                        _a), className), "aria-labelledby": ariaLabelledBy, "aria-describedby": ariaDescribedBy }), data.map(function (item, index) {
                    return (React.createElement(ChipComponent, { role: 'option', dataItem: item, size: size, key: [valueGetter(item), index].join('-'), text: textGetter(item), value: valueGetter(item) }));
                }))))));
});
var propTypes = {
    id: PropTypes.string,
    className: PropTypes.string,
    tabIndex: PropTypes.number,
    data: PropTypes.any,
    defaultData: PropTypes.arrayOf(PropTypes.any),
    onDataChange: PropTypes.func,
    value: PropTypes.oneOfType([PropTypes.any, PropTypes.arrayOf(PropTypes.any)]),
    defaultValue: PropTypes.oneOfType([PropTypes.any, PropTypes.arrayOf(PropTypes.any)]),
    onChange: PropTypes.func,
    selection: PropTypes.oneOf(['single', 'none', 'multiple']),
    textField: PropTypes.string,
    valueField: PropTypes.string,
    disabled: PropTypes.bool,
    dir: PropTypes.oneOf(['ltr', 'rtl']),
    ariaLabelledBy: PropTypes.string,
    ariaDescribedBy: PropTypes.string,
    size: PropTypes.oneOf([null, 'small', 'medium', 'large'])
};
var defaultProps = {
    chip: Chip_1.Chip,
    size: 'medium',
    disabled: false,
    defaultValue: null,
    defaultData: [],
    dir: 'ltr',
    selection: 'none',
    textField: 'text',
    valueField: 'value',
    removable: 'removable'
};
exports.ChipList.displayName = 'KendoReactChipList';
// TODO: delete casting when @types/react is updated!
exports.ChipList.propTypes = propTypes;
exports.ChipList.defaultProps = defaultProps;
