// @progress/kendo-react-treeview/dist/npm/utils/itemIdUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDecrementedItemIdAfterRemoval = exports.isItemFirstFromSiblings = exports.getShortId = exports.getAllShortIds = exports.isIdZeroLevel = exports.isIdEmptyOrZeroLevel = exports.getDirectParentId = exports.createId = exports.getFirstChildId = exports.getIdWithoutRootParentId = exports.getRootParentId = exports.getItemById = exports.SEPARATOR = exports.ZERO_LEVEL_ZERO_NODE_ID = exports.EMPTY_ID = void 0;
/**
 * @hidden
 */
exports.EMPTY_ID = '';
/**
 * @hidden
 */
exports.ZERO_LEVEL_ZERO_NODE_ID = '0';
/**
 * @hidden
 */
exports.SEPARATOR = '_';
/**
 * @hidden
 */
function getItemById(itemId, items, childrenField) {
    if (isIdZeroLevel(itemId)) {
        return items[Number(itemId)];
    }
    else {
        var rootParentItem = items[Number(getRootParentId(itemId))];
        var subItems = (rootParentItem && rootParentItem[childrenField]) || [];
        return subItems.length ?
            getItemById(getIdWithoutRootParentId(itemId), subItems, childrenField) : undefined;
    }
}
exports.getItemById = getItemById;
/**
 * @hidden
 */
function getRootParentId(itemId) {
    return isIdEmptyOrZeroLevel(itemId) ? itemId : itemId.split(exports.SEPARATOR)[0];
}
exports.getRootParentId = getRootParentId;
/**
 * @hidden
 */
function getIdWithoutRootParentId(itemId) {
    if (isIdEmptyOrZeroLevel(itemId)) {
        return itemId;
    }
    else {
        var firstSeparatorIndex = itemId.indexOf(exports.SEPARATOR);
        return itemId.substring(firstSeparatorIndex + 1);
    }
}
exports.getIdWithoutRootParentId = getIdWithoutRootParentId;
/**
 * @hidden
 */
function getFirstChildId(itemId) {
    return createId('0', itemId);
}
exports.getFirstChildId = getFirstChildId;
/**
 * @hidden
 */
function createId(childId, parentId) {
    childId = childId.toString();
    return parentId ? parentId + exports.SEPARATOR + childId : childId;
}
exports.createId = createId;
/**
 * @hidden
 */
function getDirectParentId(itemId) {
    var lastSeparatorIndex = itemId.lastIndexOf(exports.SEPARATOR);
    return lastSeparatorIndex < 0 ? exports.EMPTY_ID : itemId.substring(0, lastSeparatorIndex);
}
exports.getDirectParentId = getDirectParentId;
/**
 * @hidden
 */
function isIdEmptyOrZeroLevel(itemId) {
    return itemId === exports.EMPTY_ID || itemId.indexOf(exports.SEPARATOR) < 0;
}
exports.isIdEmptyOrZeroLevel = isIdEmptyOrZeroLevel;
/**
 * @hidden
 */
function isIdZeroLevel(itemId) {
    return itemId !== exports.EMPTY_ID && itemId.indexOf(exports.SEPARATOR) < 0;
}
exports.isIdZeroLevel = isIdZeroLevel;
/**
 * @hidden
 */
function getAllShortIds(itemId) {
    return itemId.split(exports.SEPARATOR);
}
exports.getAllShortIds = getAllShortIds;
/**
 * @hidden
 */
function getShortId(itemId) {
    var lastSeparatorIndex = itemId.lastIndexOf(exports.SEPARATOR);
    return lastSeparatorIndex < 0 ? itemId : itemId.substring(lastSeparatorIndex + 1);
}
exports.getShortId = getShortId;
/**
 * @hidden
 */
function isItemFirstFromSiblings(itemId) {
    return getShortId(itemId) === exports.ZERO_LEVEL_ZERO_NODE_ID;
}
exports.isItemFirstFromSiblings = isItemFirstFromSiblings;
/**
 * @hidden
 */
function getDecrementedItemIdAfterRemoval(removedItemId, itemId) {
    var preservedItemId = itemId;
    removedItemId = 'r' + exports.SEPARATOR + removedItemId;
    itemId = 'r' + exports.SEPARATOR + itemId;
    var itemSharedPath = getDirectParentId(removedItemId) + exports.SEPARATOR;
    if (itemId.startsWith(itemSharedPath)) {
        var itemNotSharedPath = itemId.substring(itemSharedPath.length);
        if (itemNotSharedPath) {
            var siblingShortId = getRootParentId(itemNotSharedPath);
            if (Number(getShortId(removedItemId)) < Number(siblingShortId)) {
                var decrementedId = itemSharedPath +
                    (Number(siblingShortId) - 1).toString() +
                    itemNotSharedPath.substring(siblingShortId.length);
                return getIdWithoutRootParentId(decrementedId);
            }
        }
    }
    return preservedItemId;
}
exports.getDecrementedItemIdAfterRemoval = getDecrementedItemIdAfterRemoval;
