// @progress/kendo-react-treeview/dist/npm/utils/itemUtils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addItem = exports.removeItem = exports.getAllParents = exports.areAllDirectChildrenChecked = exports.getAllDirectIndirectChildrenIds = exports.isEnabledAndAllParentsEnabled = exports.updateItem = exports.resolveItemsIds = exports.resolveItemId = exports.isItemExpandedAndWithChildren = exports.hasChildren = void 0;
var itemIdUtils_1 = require("./itemIdUtils");
var SortedPublicItemIds_1 = require("./SortedPublicItemIds");
var misc_1 = require("./misc");
/**
 * @hidden
 */
function hasChildren(item, childrenField) {
    var children = (0, misc_1.getNestedValue)(childrenField, item);
    return Boolean(children && children.length);
}
exports.hasChildren = hasChildren;
/**
 * @hidden
 */
function isItemExpandedAndWithChildren(item, fieldsSvc) {
    return fieldsSvc.expanded(item) && hasChildren(item, fieldsSvc.getChildrenField());
}
exports.isItemExpandedAndWithChildren = isItemExpandedAndWithChildren;
/**
 * @hidden
 */
function resolveItemId(publicId, idField, items, childrenField) {
    var resolvedIds = resolveItemsIds([publicId], idField, items, childrenField);
    return resolvedIds.length ? resolvedIds[0] : undefined;
}
exports.resolveItemId = resolveItemId;
/**
 * @hidden
 */
function resolveItemsIds(publicIds, idField, items, childrenField) {
    var result = [];
    var sortedIds = new SortedPublicItemIds_1.default();
    sortedIds.init(publicIds);
    resolveItemsIdsHelper(sortedIds, idField, items, itemIdUtils_1.EMPTY_ID, result, childrenField);
    return result;
}
exports.resolveItemsIds = resolveItemsIds;
/**
 * @hidden
 */
function updateItem(items, itemId, update, cloneField, childrenField) {
    var _a;
    var itemIndex = Number((0, itemIdUtils_1.getRootParentId)(itemId));
    if (itemIndex >= items.length) {
        return items;
    }
    var result = items;
    var item = items[itemIndex];
    // Clone the item before updating it.
    if (!item[cloneField]) {
        result = items.slice();
        item = result[itemIndex] = Object.assign({}, item, (_a = {}, _a[cloneField] = true, _a));
    }
    // Directly update the item.
    if ((0, itemIdUtils_1.isIdZeroLevel)(itemId)) {
        update(item);
    }
    else if (item[childrenField]) {
        item[childrenField] = updateItem(item[childrenField], (0, itemIdUtils_1.getIdWithoutRootParentId)(itemId), update, cloneField, childrenField);
    }
    return result;
}
exports.updateItem = updateItem;
/**
 * @hidden
 */
function isEnabledAndAllParentsEnabled(itemId, items, fieldsSvc) {
    var ids = (0, itemIdUtils_1.getAllShortIds)(itemId);
    var currentItems = items;
    for (var i = 0; i < ids.length; i++) {
        var currentItem = currentItems[Number(ids[i])];
        if (fieldsSvc.disabled(currentItem)) {
            return false;
        }
        else {
            currentItems = currentItem[fieldsSvc.getChildrenField()];
        }
    }
    return true;
}
exports.isEnabledAndAllParentsEnabled = isEnabledAndAllParentsEnabled;
/**
 * @hidden
 */
function getAllDirectIndirectChildrenIds(item, itemId, childrenField, idField) {
    return idField ? getChildrenPublicIds(item) : getChildrenHierarchicalIndices(item, itemId);
    function getChildrenHierarchicalIndices(startItem, startItemId) {
        var result = [];
        var children = startItem[childrenField] || [];
        for (var i = 0; i < children.length; i++) {
            var id = (0, itemIdUtils_1.createId)(i, startItemId);
            result.push(id);
            result = result.concat(getChildrenHierarchicalIndices(children[i], id));
        }
        return result;
    }
    function getChildrenPublicIds(startItem) {
        var result = [];
        var children = startItem[childrenField] || [];
        for (var i = 0; i < children.length; i++) {
            result.push((0, misc_1.getNestedValue)(idField, children[i]));
            result = result.concat(getChildrenPublicIds(children[i]));
        }
        return result;
    }
}
exports.getAllDirectIndirectChildrenIds = getAllDirectIndirectChildrenIds;
/**
 * @hidden
 */
function areAllDirectChildrenChecked(item, itemId, idField, childrenField, check) {
    var children = item[childrenField] || [];
    return children.every(function (child, index) {
        return check.indexOf(idField ? (0, misc_1.getNestedValue)(idField, child) : (0, itemIdUtils_1.createId)(index, itemId)) > -1;
    });
}
exports.areAllDirectChildrenChecked = areAllDirectChildrenChecked;
/**
 * @hidden
 */
function getAllParents(itemId, childrenField, items) {
    var result = [];
    var ids = (0, itemIdUtils_1.getAllShortIds)(itemId);
    var currentItems = items;
    for (var i = 0; i < ids.length - 1; i++) {
        if (!currentItems) {
            break;
        }
        var currentItem = currentItems[Number(ids[i])];
        result.push(currentItem);
        currentItems = currentItem[childrenField];
    }
    return result;
}
exports.getAllParents = getAllParents;
/**
 * @hidden
 */
function removeItem(itemId, childrenField, items) {
    var result = items.slice();
    if ((0, itemIdUtils_1.isIdZeroLevel)(itemId)) {
        result.splice(Number(itemId), 1);
    }
    else {
        var rootParentIndex = Number((0, itemIdUtils_1.getRootParentId)(itemId));
        var rootParent = result[rootParentIndex] = __assign({}, result[rootParentIndex]);
        rootParent[childrenField] = removeItem((0, itemIdUtils_1.getIdWithoutRootParentId)(itemId), childrenField, rootParent[childrenField]);
    }
    return result;
}
exports.removeItem = removeItem;
/**
 * @hidden
 */
function addItem(item, operation, childrenField, targetItemId, items) {
    var result = items.slice();
    if ((0, itemIdUtils_1.isIdZeroLevel)(targetItemId)) {
        if (operation === 'child') {
            var targetItem = result[Number(targetItemId)] = __assign({}, result[Number(targetItemId)]);
            if (targetItem[childrenField]) {
                targetItem[childrenField] = targetItem[childrenField].slice();
                targetItem[childrenField].push(item);
            }
            else {
                targetItem[childrenField] = [item];
            }
        }
        else {
            result.splice(Number(targetItemId) + (operation === 'after' ? 1 : 0), 0, item);
        }
    }
    else {
        var rootParentIndex = Number((0, itemIdUtils_1.getRootParentId)(targetItemId));
        var rootParent = result[rootParentIndex] = __assign({}, result[rootParentIndex]);
        rootParent[childrenField] = addItem(item, operation, childrenField, (0, itemIdUtils_1.getIdWithoutRootParentId)(targetItemId), rootParent[childrenField]);
    }
    return result;
}
exports.addItem = addItem;
function resolveItemsIdsHelper(publicItemIds, idField, items, parentItemId, resolvedIds, childrenField) {
    for (var i = 0; i < items.length; i++) {
        var item = items[i];
        var id = (0, itemIdUtils_1.createId)(i, parentItemId);
        if (publicItemIds.hasId((0, misc_1.getNestedValue)(idField, item))) {
            resolvedIds.push(id);
        }
        if (hasChildren(item, childrenField)) {
            resolveItemsIdsHelper(publicItemIds, idField, (0, misc_1.getNestedValue)(childrenField, item), id, resolvedIds, childrenField);
        }
    }
}
