// @progress/kendo-react-dialogs/dist/npm/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WindowPropsContext = exports.WindowActionsBar = exports.Window = exports.DialogPropsContext = exports.DialogActionsBar = exports.Dialog = void 0;
var Dialog_1 = require("./Dialog");
Object.defineProperty(exports, "Dialog", { enumerable: true, get: function () { return Dialog_1.Dialog; } });
Object.defineProperty(exports, "DialogPropsContext", { enumerable: true, get: function () { return Dialog_1.DialogPropsContext; } });
var DialogActionsBar_1 = require("./DialogActionsBar");
Object.defineProperty(exports, "DialogActionsBar", { enumerable: true, get: function () { return DialogActionsBar_1.DialogActionsBar; } });
var Window_1 = require("./Window");
Object.defineProperty(exports, "Window", { enumerable: true, get: function () { return Window_1.Window; } });
Object.defineProperty(exports, "WindowPropsContext", { enumerable: true, get: function () { return Window_1.WindowPropsContext; } });
var WindowActionsBar_1 = require("./WindowActionsBar");
Object.defineProperty(exports, "WindowActionsBar", { enumerable: true, get: function () { return WindowActionsBar_1.WindowActionsBar; } });
